packages/node-patches/src/fs.ts

```diff
@@ -0,0 +1,345 @@
+    const cb = args.length > 1 ? args[args.length - 1] : undefined;
+    if (cb) {
+      args[args.length - 1] = (err: Error, str: string) => {
+        if (err) return cb(err);
+        if (isOutPath(str)) {
+          const e = new Error(
+            "EINVAL: invalid argument, readlink '" + args[0] + "'"
+          );
+          //tslint:disable-next-line:no-any
+          (e as any).code = 'EINVAL';
+          // if its not supposed to be a link we have to trigger an EINVAL error.
+          cb(e);
+        } else {
+          cb(false, str);
+        }
+      };
+    }
+    origReadlink(...args);
+  };
+
+  //tslint:disable-next-line:no-any
+  fs.lstatSync = (...args: any[]) => {
+    let stats = origLstatSync(...args);
+    if (stats.isSymbolicLink() && isOutPath(origRealpathSync(args[0]))) {
+      stats = fs.statSync(...args);
+    }
+    return stats;
+  };
+
+  //tslint:disable-next-line:no-any
+  fs.realpathSync = (...args: any[]) => {
+    const str = origRealpathSync(...args);
+    if (isOutPath(str)) {
+      return path.resolve(args[0]);
+    }
+    return str;
+  };
+
+  //tslint:disable-next-line:no-any
+  fs.readlinkSync = (...args: any[]) => {
+    const str = origReadlinkSync(...args);
+    if (isOutPath(str)) {
+      const e = new Error(
+        "EINVAL: invalid argument, readlink '" + args[0] + "'"
+      );
+      //tslint:disable-next-line:no-any
+      (e as any).code = 'EINVAL';
+      throw e;
+    }
+    return str;
+  };
+
+  //tslint:disable-next-line:no-any
+  fs.readdir = (...args: any[]) => {
+    const p = path.resolve(args[0]);
+
+    const cb = args[args.length - 1];
+    if (typeof cb !== 'function') {
+      // this will likely throw callback required error.
+      return origReaddir(...args);
+    }
+
+    args[args.length - 1] = (err: Error, result: Dirent[]) => {
+      if (err) return cb(err);
+      // user requested withFileTypes
+      if (result[0] && result[0].isSymbolicLink) {
+        Promise.all(result.map((v: Dirent) => handleDirent(p, v)))
+          .then(() => {
+            cb(null, result);
+          })
+          .catch(err => {
+            cb(err);
+          });
+      } else {
+        // string array return for readdir.
+        cb(null, result);
+      }
+    };
+
+    origReaddir(...args);
+  };
+
+  //tslint:disable-next-line:no-any
+  fs.readdirSync = (...args: any[]) => {
+    const res = origReaddirSync(...args);
+    const p = path.resolve(args[0]);
+    //tslint:disable-next-line:no-any
+    res.forEach((v: Dirent | any) => {
+      handleDirentSync(p, v);
+    });
+    return res;
+  };
+
+  // i need to use this twice in bodt readdor and readdirSync. maybe in fs.Dir
+  //tslint:disable-next-line:no-any
+  function patchDirent(dirent: Dirent | any, stat: Stats | any) {
+    // add all stat is methods to Dirent instances with their result.
+    for (const i in stat) {
+      if (i.indexOf('is') === 0 && typeof stat[i] === 'function') {
+        const result = stat[i]();
```

Comment:
This could use a comment I think. It's not quite obvious why this is eagerly computed. Assuming that the original functions were using `this`, it may also change the behavior (suddenly works unbound)..?

Reply:
this class just compares properties within an object guarded by a private symbol. it might be better if i created a variable scoped object with the properties and compare them like the original class. 
i'm not sure what i can do about the works unbound at the moment. i could assign another private symbol to each class and keep the file type there but id prefer not to.

ill document it as an incompatibility.